custom_components/rrd/__init__.py

```diff
@@ -123,47 +134,95 @@ def setup(hass, config):
             _LOGGER.error(exc)
             return False
 
-    def rrd_update(event):
-        state = event.data.get("new_state")
-        if state is None:
-            return
 
-        if state.entity_id not in entities:
-            return
+    def update(database):
+        step = convert_to_seconds(database[CONF_STEP])
+        _LOGGER.debug("%s will be updated every %s seconds.", database[CONF_NAME] + ".rrd", step)
 
-        _last_changed_ts = int(datetime.timestamp(state.last_changed))
-        _state = int(state_helper.state_as_number(state))
+        rrd_filename = hass.config.path(rrd_dir, database[CONF_NAME]) + ".rrd"
 
-        ds_name, _, _ = entities[state.entity_id]
-        entities[state.entity_id] = ds_name, _last_changed_ts, _state
+        while not hassIsShuttingDown:
+            # Wait for the begining of next `step` interval.
+            now = time.time()
+            nextSavingTimestamp = ((now // step) + 1) * step
+            time.sleep(nextSavingTimestamp - now)
```

Comment:
I have hopes that we can integrate this custom component into HA.

Having a sleep() and threads is a big red flag, we will not get this accepted...

Reply:
Maybe something using https://github.com/home-assistant/core/blob/ee914366a8762f495a585e09a0bd7b0675b4011c/homeassistant/helpers/event.py#L1292

Reply:
Refactorized to async calls by `hass.loop.call_at()`.
See commit https://github.com/dgomes/rrd_recorder/pull/3/commits/c3ee4d5e2ebc7b399c3163373e35cb6afd16ad69